From 747ef469ffc9c9179ca9d76854167925b4e44346 Mon Sep 17 00:00:00 2001
From: Siddhesh Poyarekar <siddhesh@redhat.com>
Date: Sat, 15 Jun 2013 00:09:26 +0530
Subject: [PATCH] Add rtld-memset.S for x86_64

Resolves: BZ #15627

Add an assembler version of rtld-memset to avoid using SSE registers.
---
 sysdeps/x86_64/rtld-memset.S | 37 +++++++++++++++++++++++++++++++++++++
 sysdeps/x86_64/rtld-memset.c |  1 -
 4 files changed, 45 insertions(+), 2 deletions(-)
 create mode 100644 sysdeps/x86_64/rtld-memset.S
 delete mode 100644 sysdeps/x86_64/rtld-memset.c

diff --git a/sysdeps/x86_64/rtld-memset.S b/sysdeps/x86_64/rtld-memset.S
new file mode 100644
index 0000000..18b4903
--- /dev/null
+++ b/sysdeps/x86_64/rtld-memset.S
@@ -0,0 +1,37 @@
+/* memset implementation for the dynamic linker.  This is separate from the
+   libc implementation to avoid writing to SSE registers.
+   Copyright (C) 2013 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <sysdep.h>
+#include "asm-syntax.h"
+
+
+	.text
+/* void *memset (void *dest, char c, size_t count)
+   dest	 => %rdi
+   c	 => %rsi
+   count => %rdx  */
+ENTRY (memset)
+	mov	%rdx, %rcx
+	movzbl	%sil, %eax
+	mov	%rdi, %rdx
+	rep	stosb
+	mov	%rdx, %rax
+	ret
+END (memset)
+libc_hidden_builtin_def (memset)
diff --git a/sysdeps/x86_64/rtld-memset.c b/sysdeps/x86_64/rtld-memset.c
deleted file mode 100644
index 55f3835..0000000
--- a/sysdeps/x86_64/rtld-memset.c
+++ /dev/null
@@ -1 +0,0 @@
-#include <string/memset.c>
-- 
1.8.3.1

